File: ros/test_ws/src/geekbot_pkg/scripts/geekbot_link.py
#!/usr/bin/env python
import rospy
import serial
import threading
import time
from struct import pack
from collections import namedtuple
from geometry_msgs.msg import Twist
from std_msgs.msg import Bool, Int16

ser = serial.Serial('/dev/geekbot/arduino', 115200)
ser.timeout = None
time.sleep(.5)
print("Serial port opened.")
pub = rospy.Publisher('ir_mm', Int16, queue_size=5)

def map_short(num):
    temp = num * 32767
    if temp > 32767:
        return 32767
    elif temp < -32767:
        return -32767

    return int(num * 32767)
    
def pack_short(num):
    return pack("h", int(num))

def get_speeds(linear, angular):
    motors = namedtuple("motors",'left right')
    motors.right = -linear - angular
    motors.left  = -linear + angular
    return motors

def send_cmd(flag, data):
    ser.write(chr(flag))
    ser.write(pack_short(map_short(data)))

def left_motor(data):
    send_cmd(0x36, data)

def right_motor(data): 
    send_cmd(0x35, data)

def cmd_vel_callback(data):
    if data.linear.x or data.angular.z:
        speeds = get_speeds(data.linear.x, data.angular.z)
        left_motor(speeds.left)
        right_motor(speeds.right)
    elif data.linear.y or data.linear.z:
        left_motor(data.linear.y)
        right_motor(data.linear.z)

def lights_callback(thing):
    if thing.data == True:
        send_cmd(0x30, 1)
        print("Lights callback: on")
    elif thing.data == False:
        send_cmd(0x30, 0)
        print("Lights callback: off")

def buzzer_callback(thing):
    if thing.data == True:
        send_cmd(0x29, 1)
        print("Buzzer callback: on")
    elif thing.data == False:
        send_cmd(0x29, 0)
        print("Buzzer callback: off")

def ir_update(yes):
    if yes:
        send_cmd(0x27,1)
        mm = ser.read()*256
        #print(mm)
	mm = ser.read()+mm
	print(mm)
        pub.publish(mm) 
    else:
        yes = not yes
    return yes 
ir_update.counter = 0

if __name__ == '__main__':
    rospy.init_node('geekbot_link', anonymous=False)
    r = rospy.Rate(60)
    rospy.Subscriber("lights", Bool, lights_callback)
    rospy.Subscriber("cmd_vel", Twist, cmd_vel_callback)
    rospy.Subscriber("buzzer", Bool, buzzer_callback)
    update = False
    while not rospy.is_shutdown():
        r.sleep()
	update = ir_update(update)

    

File: ros/test_ws/src/geekbot_pkg/src/geekbot_link.cpp
#include <string>
#include <iostream>
#include <cstdio>
#include <unistd.h>
#include <sys/time.h>
#include <thread>

#include "easy_serial.h"
#include <ros/ros.h>
#include <geometry_msgs/Twist.h>
#include <std_msgs/Int16.h>
#include <std_msgs/Bool.h>

#define BAUDRATE 57600
#define TIMEOUT 50
#define READ_DELAY 10
#define IR_POLL 100

using std::string;
using std::cout;
using std::vector;
using std::endl;

struct motor_speeds
{
	short left;
	short right;
} motors;

volatile int ir_distance; 
volatile int main_thread_dead = 0;
volatile int main_thread;

int arduino = 0;

short map_short(float num)
{
    short temp = num * 32767;
    if (temp > 32767)
        return 32767;
    else if (temp < -32767)
        return -32767;

    return temp;
}

bool send_cmd(unsigned char flag, short data)
{
	unsigned char flag_temp = flag;
	short data_temp = data;
	int n  = write(arduino, &flag_temp, 1);
	n += write(arduino, &data_temp, sizeof(data_temp));
	if(n > 1 + sizeof(data));
		return 0;
	return 1;
}

bool set_left(short data){
    return send_cmd(0x36, data);
}

bool set_right(short data){
    return send_cmd(0x35, data);
}

bool lights_callback(std_msgs::Bool thing){
    if (thing.data){
        send_cmd(0x30, 1);
        cout << "Lights callback: on\n";
        }
    else{
        send_cmd(0x30, 0);
        cout << "Lights callback: off\n";
	}
}

bool buzzer_callback(std_msgs::Bool thing){
    if (thing.data){
        send_cmd(0x29, 1);
        cout << "Buzzer callback: on\n";
    }
    else{
        send_cmd(0x29, 0);
        cout << "Buzzer callback: off\n";
    }
}

bool ir_update(){
    //if(send_cmd(0x27, 1))
//	cout << "IR update error!" <<endl;
    /*unsigned char temp = 0x27;
    write(arduino, &temp, 1);
    write(arduino, &temp, 1);
    write(arduino, &temp, 1);*/
    if(send_cmd(0x27, 1))
        cout << "IR update send error." << endl;
    unsigned char high_byte = 0;
    unsigned char low_byte = 0;
    usleep(READ_DELAY*1000);
    read(arduino, &high_byte, 1);
    read(arduino, &low_byte, 1);
    short ir_temp = (high_byte << 8) | low_byte;
    if(ir_temp){
        ir_distance = ir_temp;
        return 0;
    }
    return 1;
}

void ir_func(int main_thread)
{
	while(!main_thread_dead){
	if(!ir_update())
	{
		cout << "ir_distance = " << ir_distance << endl;
	}
	usleep(IR_POLL * 1000);
	}
}

struct motor_speeds get_speeds(float linear, float angular)
{
	struct motor_speeds temp;
	temp.right = -linear - angular;
	temp.left  = -linear + angular;
	return temp;
}


bool cmd_vel_callback(geometry_msgs::Twist data){
    if (data.linear.x || data.angular.z){
        motors = get_speeds(data.linear.x, data.angular.z);
        set_left(motors.left);
        set_right(motors.right);
    }
    else if (data.linear.y || data.linear.z){
        set_left(data.linear.y);
        set_right(data.linear.z);
    }
}

int main(int argc, char **argv)
{
	arduino = serial_port_init("/dev/geekbot/arduino", BAUDRATE);
	//main_thread = std::this_thread::get_id();

	if(arduino < 1)
	{
		cout << "Can't open port, exiting...." << endl;
		main_thread_dead = 1;
		return 1;
	}
	//clear_port(arduino);
	unsigned char temp = 0;
	unsigned char handshake = 0x77;
	sleep(2);
	while(temp != handshake){	
        write(arduino, &handshake, 1);
		int n = read(arduino,&temp,1);
		usleep(10*1000);
		printf("Looking for handshake\n");
		if(n)printf("Handshake received: %0x\n", temp);}
    clear_port(arduino);
	std::thread ir_thread (ir_func, main_thread);
	sleep(10);
	main_thread_dead = 1;
	ir_thread.join();
	return 0;
}	
